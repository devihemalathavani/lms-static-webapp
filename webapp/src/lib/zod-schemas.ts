import { z } from "zod";

export const EmailSchema = z
  .string()
  .email()
  .trim()
  .transform((v) => v.toLowerCase());
