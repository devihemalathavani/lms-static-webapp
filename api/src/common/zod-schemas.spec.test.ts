import { describe, expect, it } from "vitest";
import { ZodSchema } from "zod";
import { EmailSchema, IdSchema } from "./zod-schemas";

const errorSnapshot = Symbol("zod error snapshot");
const valueSnapshot = Symbol("zod value snapshot");

const testCases: Array<[ZodSchema, unknown, unknown, string?]> = [
  [IdSchema, undefined, errorSnapshot, "Should not accept undefined value"],
  [IdSchema, null, errorSnapshot, "Should not accept null value"],
  [IdSchema, "ramesh", errorSnapshot, "Should not accept non number values"],
  [IdSchema, {}, errorSnapshot, "Should not accept non number values"],
  [IdSchema, "43", 43, "Should accept number strings"],
  [IdSchema, 32, 32, "shoulud accept numbers"],
  [EmailSchema, "abc@example.com", "abc@example.com", "should accept emails"],
  [
    EmailSchema,
    "abxample.com",
    errorSnapshot,
    "should not accept emails without @",
  ],
];

describe("Common zod schemas", () => {
  it.each(testCases)("%# %s", (schema, data, result) => {
    if (result === errorSnapshot) {
      expect(() => schema.parse(data)).toThrowErrorMatchingSnapshot();
      return;
    }

    if (result === valueSnapshot) {
      expect(schema.parse(data)).toMatchSnapshot();
      return;
    }

    expect(schema.parse(data)).toEqual(result);
  });
});
