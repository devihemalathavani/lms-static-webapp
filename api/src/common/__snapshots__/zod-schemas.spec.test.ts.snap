// Vitest Snapshot v1

exports[`Common zod schemas > 0 ZodEffects {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional]
} 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Common zod schemas > 1 ZodEffects {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional]
} 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Common zod schemas > 2 ZodEffects {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional]
} 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Common zod schemas > 3 ZodEffects {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional]
} 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Common zod schemas > 7 ZodEffects {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional]
} 1`] = `
"[
  {
    \\"validation\\": \\"email\\",
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid email\\",
    \\"path\\": []
  }
]"
`;

exports[`Common zod schemas > 7 ZodString {
  spa: [Function: bound safeParseAsync] AsyncFunction,
  superRefine: [Function: bound _refinement],
  _def: [Object],
  parse: [Function: bound parse],
  safeParse: [Function: bound safeParse],
  parseAsync: [Function: bound parseAsync] AsyncFunction,
  safeParseAsync: [Function: bound safeParseAsync] AsyncFunction,
  refine: [Function: bound refine],
  refinement: [Function: bound refinement],
  optional: [Function: bound optional],
  nullable: [Function: bound nullable],
  nullish: [Function: bound nullish],
  array: [Function: bound array],
  promise: [Function: bound promise],
  or: [Function: bound or],
  and: [Function: bound and],
  transform: [Function: bound transform],
  default: [Function: bound default],
  describe: [Function: bound describe],
  isNullable: [Function: bound isNullable],
  isOptional: [Function: bound isOptional],
  _regex: [Function (anonymous)],
  nonempty: [Function (anonymous)],
  trim: [Function (anonymous)]
} 1`] = `
"[
  {
    \\"validation\\": \\"email\\",
    \\"code\\": \\"invalid_string\\",
    \\"message\\": \\"Invalid email\\",
    \\"path\\": []
  }
]"
`;
